server/src/main/java/org/elasticsearch/rest/MethodHandlers.java

```diff
@@ -28,12 +28,7 @@ final class MethodHandlers {
 
     MethodHandlers(String path) {
         this.path = path;
-
-        // by setting the loadFactor to 1, these maps are resized only when they *must* be, and the vast majority of these
-        // maps contain only 1 or 2 entries anyway, so most of these maps are never resized at all and waste only 1 or 0
-        // array references, while those few that contain 3 or 4 elements will have been resized just once and will still
-        // waste only 1 or 0 array references
-        this.methodHandlers = new HashMap<>(2, 1);
+        this.methodHandlers = new EnumMap<>(RestRequest.Method.class);
```

Comment:
Will this use more memory than a default HashMap with only one entry?

Reply:
Using 2-values enum `RestApiVersion` instead of 9-values `RestRequest.Method` does not change the estimates a bit.
But the values array in `EnumMap` has the size of the enum cardinality! I will have a closer look.

Reply:
Not taking into account "technical" data like refCount, vTable pointer, etc. and assuming pointer size = 4 bytes here are the estimates:

`new EnumMap<>(RestRequest.Method.class).put(...)`
- keyType = 4
- keyUniverse = 4 (array is shared)
- vals = Object[9] = 4 + 4 + 9*4
- size = 4
- entrySet = 4 (stateless)
- keySet = 4 (null)
- values = 4 (null)

TOTAL: 68 bytes
does not change when more values are added
for `new EnumMap<>(RestApiVersion.class).put(...)` => 40 bytes because vals = Object[2]


`new HashMap<>(2, 1).put(...)`
- table = Node[2] = 4 + 4 + 2*4
    table[1]: Node
    - hash = 4
    - key = 4
    - value = 4
    - next (null) = 4
- entrySet = 4 (stateless)
- size = 4
- modCount = 4
- threshold = 4
- loadFactor = 4
- keySet = 4 (null)
- values = 4 (null)

TOTAL: 60 bytes
after adding 2nd entry: 76 bytes (+ 1 Node)
after adding 3rd and 4th entries: 116 (+ 2 Node, Node[2] => Node[4])
for `new HashMap<>(1, 1).put(...)` => 56 bytes because table = Node[1]

So `HashMap` loses in every scenario except for the 1-entry case with `RestRequest.Method` keys.
In reality, this 8 bytes advantage will be subsumed by the extra Node object overhead.

Reply:
OK, that looks sensible to me